File: plugins/cache/cc-query-dev/cc-query/0.6.1/zig/src/formatter.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const database = @import("database.zig");

/// Format result set as TSV (tab-separated values)
pub fn formatTsv(allocator: Allocator, result: *const database.ResultSet) ![]const u8 {
    if (result.columns.len == 0) return try allocator.dupe(u8, "");

    var output: std.ArrayListUnmanaged(u8) = .{};
    errdefer output.deinit(allocator);

    // Header row
    for (result.columns, 0..) |col, i| {
        if (i > 0) try output.append(allocator, '\t');
        try output.appendSlice(allocator, col);
    }
    try output.append(allocator, '\n');

    // Data rows
    for (result.rows) |row| {
        for (row, 0..) |cell, i| {
            if (i > 0) try output.append(allocator, '\t');
            try output.appendSlice(allocator, cell);
        }
        try output.append(allocator, '\n');
    }

    return try output.toOwnedSlice(allocator);
}

/// Format result set as a table with box-drawing characters
pub fn formatTable(allocator: Allocator, result: *const database.ResultSet) ![]const u8 {
    if (result.columns.len == 0) return try allocator.dupe(u8, "");

    var output: std.ArrayListUnmanaged(u8) = .{};
    errdefer output.deinit(allocator);

    const col_count = result.columns.len;

    // Calculate column widths
    var widths = try allocator.alloc(usize, col_count);
    defer allocator.free(widths);

    for (result.columns, 0..) |col, i| {
        widths[i] = col.len;
    }

    for (result.rows) |row| {
        for (row, 0..) |cell, i| {
            if (cell.len > widths[i]) widths[i] = cell.len;
        }
    }

    // Top border: ┌──┬──┐
    try output.appendSlice(allocator, "┌");
    for (widths, 0..) |w, i| {
        if (i > 0) try output.appendSlice(allocator, "┬");
        try appendRepeated(&output, allocator, "─", w + 2);
    }
    try output.appendSlice(allocator, "┐\n");

    // Header row: │ col1 │ col2 │
    try output.appendSlice(allocator, "│ ");
    for (result.columns, 0..) |col, i| {
        if (i > 0) try output.appendSlice(allocator, " │ ");
        try output.appendSlice(allocator, col);
        try appendSpaces(&output, allocator, widths[i] - col.len);
    }
    try output.appendSlice(allocator, " │\n");

    // Header separator: ├──┼──┤
    try output.appendSlice(allocator, "├");
    for (widths, 0..) |w, i| {
        if (i > 0) try output.appendSlice(allocator, "┼");
        try appendRepeated(&output, allocator, "─", w + 2);
    }
    try output.appendSlice(allocator, "┤\n");

    // Data rows
    for (result.rows) |row| {
        try output.appendSlice(allocator, "│ ");
        for (row, 0..) |cell, i| {
            if (i > 0) try output.appendSlice(allocator, " │ ");
            try output.appendSlice(allocator, cell);
            try appendSpaces(&output, allocator, widths[i] - cell.len);
        }
        try output.appendSlice(allocator, " │\n");
    }

    // Bottom border: └──┴──┘
    try output.appendSlice(allocator, "└");
    for (widths, 0..) |w, i| {
        if (i > 0) try output.appendSlice(allocator, "┴");
        try appendRepeated(&output, allocator, "─", w + 2);
    }
    try output.appendSlice(allocator, "┘\n");

    // Row count
    const row_word = if (result.rows.len == 1) "row" else "rows";
    const footer = try std.fmt.allocPrint(allocator, "({d} {s})\n", .{ result.rows.len, row_word });
    defer allocator.free(footer);
    try output.appendSlice(allocator, footer);

    return try output.toOwnedSlice(allocator);
}

fn appendSpaces(output: *std.ArrayListUnmanaged(u8), allocator: Allocator, count: usize) !void {
    for (0..count) |_| {
        try output.append(allocator, ' ');
    }
}

fn appendRepeated(output: *std.ArrayListUnmanaged(u8), allocator: Allocator, s: []const u8, count: usize) !void {
    for (0..count) |_| {
        try output.appendSlice(allocator, s);
    }
}

test "formatTsv basic" {
    const allocator = std.testing.allocator;

    // Create a fake ResultSet
    var columns = try allocator.alloc([]const u8, 2);
    columns[0] = try allocator.dupe(u8, "id");
    columns[1] = try allocator.dupe(u8, "name");

    var rows = try allocator.alloc([][]const u8, 2);

    rows[0] = try allocator.alloc([]const u8, 2);
    rows[0][0] = try allocator.dupe(u8, "1");
    rows[0][1] = try allocator.dupe(u8, "hello");

    rows[1] = try allocator.alloc([]const u8, 2);
    rows[1][0] = try allocator.dupe(u8, "2");
    rows[1][1] = try allocator.dupe(u8, "world");

    var result = database.ResultSet{
        .columns = columns,
        .rows = rows,
        .allocator = allocator,
    };
    defer result.deinit();

    const tsv = try formatTsv(allocator, &result);
    defer allocator.free(tsv);

    try std.testing.expectEqualStrings("id\tname\n1\thello\n2\tworld\n", tsv);
}

test "formatTable basic" {
    const allocator = std.testing.allocator;

    var columns = try allocator.alloc([]const u8, 2);
    columns[0] = try allocator.dupe(u8, "id");
    columns[1] = try allocator.dupe(u8, "name");

    var rows = try allocator.alloc([][]const u8, 1);
    rows[0] = try allocator.alloc([]const u8, 2);
    rows[0][0] = try allocator.dupe(u8, "1");
    rows[0][1] = try allocator.dupe(u8, "hello");

    var result = database.ResultSet{
        .columns = columns,
        .rows = rows,
        .allocator = allocator,
    };
    defer result.deinit();

    const table = try formatTable(allocator, &result);
    defer allocator.free(table);

    // Check it has box-drawing characters and row count
    try std.testing.expect(std.mem.indexOf(u8, table, "┌") != null);
    try std.testing.expect(std.mem.indexOf(u8, table, "│") != null);
    try std.testing.expect(std.mem.indexOf(u8, table, "(1 row)") != null);
}

File: plugins/cache/cc-query-dev/cc-query/0.6.1/ccq/src/formatter.rs
//! Output formatting for query results.

use chrono::{TimeZone, Utc};
use duckdb::types::{TimeUnit, ValueRef};
use std::fmt::{self, Display, Formatter};

/// Wrapper for displaying `ValueRef` without allocation for text.
pub struct DisplayValueRef<'a>(pub &'a ValueRef<'a>);

impl Display for DisplayValueRef<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self.0 {
            ValueRef::Null => write!(f, "NULL"),
            ValueRef::Boolean(b) => write!(f, "{b}"),
            ValueRef::TinyInt(n) => write!(f, "{n}"),
            ValueRef::SmallInt(n) => write!(f, "{n}"),
            ValueRef::Int(n) => write!(f, "{n}"),
            ValueRef::BigInt(n) => write!(f, "{n}"),
            ValueRef::HugeInt(n) => write!(f, "{n}"),
            ValueRef::Float(n) => write!(f, "{n}"),
            ValueRef::Double(n) => write!(f, "{n}"),
            ValueRef::Text(bytes) => {
                let s = std::str::from_utf8(bytes).unwrap_or("<invalid utf8>");
                write!(f, "{s}")
            }
            ValueRef::Timestamp(unit, val) => {
                write!(f, "{}", format_timestamp(*unit, *val))
            }
            ValueRef::Date32(days) => write!(f, "{}", format_date(*days)),
            ValueRef::Blob(bytes) => write!(f, "<{} bytes>", bytes.len()),
            _ => write!(f, "{:?}", self.0),
        }
    }
}

/// Format a timestamp to match Node.js output: "YYYY-MM-DD HH:MM:SS.mmm"
fn format_timestamp(unit: TimeUnit, value: i64) -> String {
    let micros = match unit {
        TimeUnit::Second => value * 1_000_000,
        TimeUnit::Millisecond => value * 1_000,
        TimeUnit::Microsecond => value,
        TimeUnit::Nanosecond => value / 1_000,
    };

    let Some(dt) = Utc.timestamp_micros(micros).single() else {
        return "INVALID_TIMESTAMP".into();
    };
    dt.format("%Y-%m-%d %H:%M:%S%.3f").to_string()
}

/// Format a date (days since Unix epoch) to "YYYY-MM-DD"
fn format_date(days: i32) -> String {
    // Unix epoch is 1970-01-01, which is day 719,163 in the CE calendar
    let Some(d) = chrono::NaiveDate::from_num_days_from_ce_opt(days + 719_163) else {
        return "INVALID_DATE".into();
    };
    d.format("%Y-%m-%d").to_string()
}

/// Format results as a table with Unicode box-drawing characters.
///
/// Format matches Node.js exactly:
/// ```text
/// ┌──────────┬───────┐
/// │ column1  │ col2  │
/// ├──────────┼───────┤
/// │ value1   │ val2  │
/// └──────────┴───────┘
/// (N rows)
/// ```
pub fn format_table(columns: &[String], rows: &[Vec<String>]) -> String {
    if rows.is_empty() {
        // Special case: header only with "(0 rows)"
        return format!("{}\n(0 rows)", columns.join(" | "));
    }

    // Calculate column widths (max of header and data)
    let widths: Vec<usize> = columns
        .iter()
        .enumerate()
        .map(|(i, name)| {
            let max_data = rows
                .iter()
                .map(|r| r.get(i).map_or(0, String::len))
                .max()
                .unwrap_or(0);
            name.len().max(max_data)
        })
        .collect();

    let mut lines = Vec::new();

    // Top border: ┌─────┬─────┐
    let top = format!(
        "┌{}┐",
        widths
            .iter()
            .map(|w| "─".repeat(w + 2))
            .collect::<Vec<_>>()
            .join("┬")
    );
    lines.push(top);

    // Header row: │ col1  │ col2  │
    let header = columns
        .iter()
        .enumerate()
        .map(|(i, name)| format!("{:width$}", name, width = widths[i]))
        .collect::<Vec<_>>()
        .join(" │ ");
    lines.push(format!("│ {header} │"));

    // Header separator: ├─────┼─────┤
    let sep = format!(
        "├{}┤",
        widths
            .iter()
            .map(|w| "─".repeat(w + 2))
            .collect::<Vec<_>>()
            .join("┼")
    );
    lines.push(sep);

    // Data rows: │ val1  │ val2  │
    for row in rows {
        let row_str = row
            .iter()
            .enumerate()
            .map(|(i, val)| format!("{:width$}", val, width = widths[i]))
            .collect::<Vec<_>>()
            .join(" │ ");
        lines.push(format!("│ {row_str} │"));
    }

    // Bottom border: └─────┴─────┘
    let bottom = format!(
        "└{}┘",
        widths
            .iter()
            .map(|w| "─".repeat(w + 2))
            .collect::<Vec<_>>()
            .join("┴")
    );
    lines.push(bottom);

    // Row count
    let row_word = if rows.len() == 1 { "row" } else { "rows" };
    lines.push(format!("({} {row_word})", rows.len()));

    lines.join("\n")
}

/// Format results as tab-separated values.
pub fn format_tsv(columns: &[String], rows: &[Vec<String>]) -> String {
    let mut lines = Vec::with_capacity(rows.len() + 1);
    lines.push(columns.join("\t"));
    for row in rows {
        lines.push(row.join("\t"));
    }
    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_tsv() {
        let columns = vec!["a".to_string(), "b".to_string()];
        let rows = vec![
            vec!["1".to_string(), "2".to_string()],
            vec!["3".to_string(), "4".to_string()],
        ];
        assert_eq!(format_tsv(&columns, &rows), "a\tb\n1\t2\n3\t4");
    }

    #[test]
    fn test_format_table_empty() {
        let columns = vec!["col1".to_string(), "col2".to_string()];
        let rows: Vec<Vec<String>> = vec![];
        assert_eq!(format_table(&columns, &rows), "col1 | col2\n(0 rows)");
    }

    #[test]
    fn test_format_table_with_data() {
        let columns = vec!["a".to_string(), "b".to_string()];
        let rows = vec![vec!["1".to_string(), "2".to_string()]];
        let result = format_table(&columns, &rows);
        assert!(result.contains("┌"));
        assert!(result.contains("(1 row)"));
    }
}
